Add potential interactions in 4C

davidrudlstorfer opened the pull request:
In this PR I added the following things:

- A new top level directory `FourC` (I learned that Python modules are not allowed to start with a number, e.g., `4C`) which now includes the `beam_potential.py` file for potential interactions.

- Therein, I've added a new class `BeamPotential` which helps with the header and runtime output definition as well as the boundary conditions for potential interactions

- I also added your proposed code snippet to the utilities of MeshPy, as this is a very useful tool.

The final structure is probably upon discussion, therefore, I've waited to add a test (as this only tests for simple header changes this is not yet important)

Finally I now use this code:

<details>
  <summary>Click me</summary>

  ```python
import numpy as np

from meshpy import (
    InputFile,
    GeometrySet,
    Function,
    MaterialReissner,
    Beam3rHerm2Line3,
    BoundaryCondition,
)

from meshpy.conf import mpy
from meshpy.FourC.beam_potential import BeamPotential
from meshpy.mesh_creation_functions.beam_basic_geometry import create_beam_mesh_helix
from meshpy.utility import find_node_on_plane


def main():
    """Minimal example of multiple helical beams with potential interactions."""

    ### define input file
    input_file = InputFile()

    ### define material
    mat = MaterialReissner(youngs_modulus=1000, radius=0.5, shear_correction=1.0)

    ### determine quantities of helice
    # define radii of differing helical segments
    radii = np.array([0, 3, 6, 9, 12, 15, 18])

    # determine twist angles if helical segments
    outermost_twist_angle = np.deg2rad(45)
    twist_angles = []
    for radius in radii:
        if np.isclose(radius, 0.0):
            twist_angles.append(np.pi / 2)
            continue
        twist_angles.append(
            np.arctan((radii[-1] / radius) * np.tan(outermost_twist_angle))
        )

    ### Define functions for line charge density
    fun_1 = Function("TESTFUNCTION 123")
    fun_1.n_global = 1
    fun_2 = Function("TESTFUNCTION 234")
    fun_2.n_global = 2

    ### Define beam potential
    beampotential = BeamPotential(
        input_file,
        pot_law_prefactor=[-1.0e-3, 12.45e-8],
        pot_law_exponent=[6.0, 12.0],
        pot_law_line_charge_density=[1.0, 2.0],
        pot_law_line_charge_density_funcs=[fun_1, fun_2],
    )

    ### Add header and runtime output to .dat file

    beampotential.add_header(
        cutoff_radius=10.0,
        evaluation_strategy="SingleLengthSpecific_SmallSepApprox_Simple",
        potential_reduction_length=15.0,
    )
    beampotential.add_runtime_output()

    ### Create helices and add potential charge condition to each
    for radius, twist_angle in zip(radii, twist_angles):

        helix_set = create_beam_mesh_helix(
            input_file,
            Beam3rHerm2Line3,
            mat,
            [0, 0, 1],  # axis
            [0, 0, 0],  # axis_point
            [radius, 0, radius],  # start_point
            twist_angle,  # twist_angle
            height_helix=80,
            n_el=20,
            warning_straight_line=False,
        )

        beampotential.add_potential_charge_condition(geometry_set=helix_set["line"])

    ### Add boundary condition to bottom and top node
    input_file.add(
        BoundaryCondition(
            GeometrySet(
                input_file.get_nodes_by_function(
                    find_node_on_plane, normal=[0, 0, 1], origin_distance=0.0, tol=0.1
                )
            ),
            "NUMDOF 9 ONOFF 1 1 1 1 1 1 0 0 0",
            bc_type=mpy.bc.dirichlet,
        )
    )

    input_file.add(
        BoundaryCondition(
            GeometrySet(
                input_file.get_nodes_by_function(
                    find_node_on_plane, normal=[0, 0, 1], origin_distance=98.0, tol=0.1
                )
            ),
            "NUMDOF 9 ONOFF 1 1 1 1 1 1 0 0 0",
            bc_type=mpy.bc.dirichlet,
        )
    )

    input_file.write_input_file("_test_output/test.dat", header=False)
    input_file.write_vtk("test", "_test_output/")


if __name__ == "__main__":

    main()

  ```
</details>

to create the following perfectly fine input file (`.txt` instead of `.dat` due to Github not supporting `.dat` files)
[test_potential_interaction.txt](https://github.com/imcs-compsim/meshpy/files/15200887/test_potential_interaction.txt)

I would be more than happy to get your feedback regarding these changes 😊

____________________

Closes #72

davidrudlstorfer commented:
One thing I am unsure about is how to add a global (or get) a global ID (number) for a defined function?

I specifically set it with

`fun_1.n_global = 1`

isteinbrecher commented on meshpy/FourC/__init__.py:
```diff
@@ -0,0 +1,38 @@
+# SOFTWARE.
+# -----------------------------------------------------------------------------
+"""
+This module defines classes and functions to create a 4C beam input file.
```
```suggestion
This module defines classes and functions for the interaction of MeshPy with 4C.
```

davidrudlstorfer replied:
Added in latest commit.

isteinbrecher replied:
Is it possible that you did not push the new commit yet?

davidrudlstorfer replied:
Yea - my bad. I am currently experiencing a few problems with the unit tests. Once these are resolved locally I will push everything

isteinbrecher commented on meshpy/FourC/beam_potential.py:
```diff
@@ -0,0 +1,256 @@
+from ..boundary_condition import BoundaryCondition
+
+
+class BeamPotential:
```
Maybe add a short description on what the purpose of the class is, how it should be used.

davidrudlstorfer replied:
Added  a short description in the last commit.

isteinbrecher commented on meshpy/FourC/beam_potential.py:
```diff
@@ -0,0 +1,256 @@
+            Number of Gauss points to be used per integration segment.
+        potential_reduction_length: float
+            Potential is smoothly decreased within this length when using the
+            singlelengthsspecific approach to enable an axial pull off force.
```
Is `singlelengthsspecific` a well know word in the beam potential framework?

davidrudlstorfer replied:
I would say so, a big differentiation between single length specific (SBIP) and double length specific (SSIP) methods is possible.

isteinbrecher replied:
Ok, can you then change the documentation to something like:
```
Potential is smoothly decreased within this length when using the
single length specific (SBIP) approach to enable an axial pull off force.
```
The compound word `singlelengthsspecific` looks weird / like a typo in the documentation.

isteinbrecher commented on meshpy/FourC/beam_potential.py:
```diff
@@ -0,0 +1,256 @@
+        regularization_separation=None,
+        integration_segments=1,
+        gauss_points=10,
+        potential_reduction_length=-1,
```
Does `-1` make sense here, or ist ist just a default "empty" value? In case of the latter, I would make the default value `None`.

davidrudlstorfer replied:
`-1` disables the potential reduction length, therefore, I would keep it disabled as a default value.

isteinbrecher commented on meshpy/FourC/beam_potential.py:
```diff
@@ -0,0 +1,256 @@
+            if func != "none":
+                self.input_file.add(func)
+
+            bc = BoundaryCondition(
+                geometry_set,
+                f"POTLAW {i+1} VAL {line_charge} FUNCT {str(func)}",
+                bc_type="DESIGN LINE BEAM POTENTIAL CHARGE CONDITIONS",
+            )
```
```suggestion
            bc = BoundaryCondition(
                geometry_set,
                f"POTLAW {i+1} VAL {line_charge} FUNCT {{}}",
                bc_type="DESIGN LINE BEAM POTENTIAL CHARGE CONDITIONS",
                format_replacement=[func],
            )
```
The `BoundaryCondition` object has the functionality to evaluate certain string replacements only when the actual line for the input file is created. In other words, the `InputFile` can create the ordering and indexing of the functions and then the boundary condition automatically inserts the correct function ID. This mechanism is not perfect, but works quire well for the replacement of function IDs in BC definitions. This should fix the issue you raised in https://github.com/imcs-compsim/meshpy/pull/74#issuecomment-2093084025. I think in the case of `func="none"` this should also work, but please make a quick test to verify this.

davidrudlstorfer replied:
Thanks for this suggestion! It works like a charm, both for defined functions, as well as the `none` case.

isteinbrecher commented on meshpy/utility.py:
```diff
@@ -328,3 +328,23 @@ def get_min_max_nodes(nodes, *, middle_nodes=False):
     return geometry
+
+
+def find_node_on_plane(node, *, normal=None, origin_distance=None, tol=mpy.eps_pos):
```
Maybe change the name to `is_node_on_plane` or  `query_node_on_plane`? Since this function does not actually search/find anything.

Also, lets add an second optional parameter `point_on_plane` (mutually exclusive with `origin_distance`) where one can give a point on the plane to define it, I assume in some cases it can be beneficial yo have this option as well.

davidrudlstorfer replied:
Added the second option in the latest commit.

isteinbrecher commented on pyproject.toml:
```diff
@@ -5,6 +5,7 @@ requires = ["setuptools", "Cython", "numpy", "wheel"]
 [tool.setuptools]
 packages = [
   "meshpy",
+  "meshpy.FourC",
```
I am not a fan of the spelling `FourC`, I would go with `four_c` to align with the spelling of other submodules in MeshPy.

isteinbrecher replied:
Also for  testing I think you have to add the submodule to the `pyproject.toml` file.

davidrudlstorfer replied:
Changed it to your suggestion of `four_c`, now the code is consistent with the remaining parts.

Maybe you could briefly elaborate what you mean by adding the submodule to the `pyproject.toml` file? Unfortunately, I do not understand it

isteinbrecher replied:
My bad, you already added the submodule correctly here.

isteinbrecher commented on meshpy/utility.py:
```diff
@@ -328,3 +328,23 @@ def get_min_max_nodes(nodes, *, middle_nodes=False):
+
+def find_node_on_plane(node, *, normal=None, origin_distance=None, tol=mpy.eps_pos):
+    """
+    Return the nodes lie on the plane defined by normal and origin_distance.
```
```suggestion
    Query if a node lies on a plane defined by a point_on_plane or the origin distance.
```

davidrudlstorfer replied:
Changed in latest commit.

isteinbrecher commented on meshpy/utility.py:
```diff
@@ -328,3 +328,23 @@ def get_min_max_nodes(nodes, *, middle_nodes=False):
+    origin_distance: float
+        Distance between origin and defined plane.
+    tol: float
+        Tolerance of evaluation if point coincides with plane
```
```
Return:
True if the point lies on the plane, False otherwise.
```

davidrudlstorfer replied:
Added

isteinbrecher reviewed (commented):
Thanks @davidrudlstorfer for the contribution. I left some comments. Also can you please add a test case to cover your changes? One (or how many you need) for the beam potential interactions and one where we check the `nodes_on_plane` function?

isteinbrecher commented:
@davidrudlstorfer in general the script looks good to me, one comment:
```python
input_file.write_input_file("_test_output/test.dat", header=False)
```
In testing we always turn the header off, but in practice I would keep it on, as it adds some information to your input file that can later be used to help to identify the state of the script when the file was created.

davidrudlstorfer commented:
@isteinbrecher thanks for your review! I've added all of your recommendations to latest changes.  Also thanks for the help with the function numbering within the `.dat` file.

Regarding the tests:

I took the liberty and renamed `testing_utility.py` into `utilities.py` and created a `testing_utility.py` for the newly added node testing function. This was done to adhere with the current file naming scheme of the test files.

isteinbrecher commented:
@davidrudlstorfer everything looks good to me, if you think this is ready, we can merge this.

isteinbrecher commented:
Can you click on the merge button or is this for member of the group `imcs-compsim` only?

davidrudlstorfer commented:
Unfortunately I am not authorized to push the big button

isteinbrecher commented:
I invited you to the project with `write` acces, does it work now?

davidrudlstorfer commented:
@isteinbrecher thanks for the invitation, unfortunately it still does not work

<img width="913" alt="Screenshot 2024-05-06 at 18 52 47" src="https://github.com/imcs-compsim/meshpy/assets/106265296/6d59f4da-5e58-413b-b811-8ae20431bdf4">

isteinbrecher commented:
How about now?

davidrudlstorfer commented:
Yes - worked like a charm. Thanks for your help and review regarding this PR! This concludes the first steps for me regarding MeshPy. I am now going to include it within my own repo for the creation of input files. I am going to contribute further to MeshPy depending on different upcoming needs :smile:

isteinbrecher commented:
Thanks for your contributions and looking forward to the next ones! I might come up to you for some small improvements cleanup in the future 😉
